### PR TITLE
Issue #503 - Add optional client_id to OAuth2

### PR DESCRIPTION
**Pull request by bbooth**
This is a new pull request after doing a rebase on the ember-cli-addon.  I will close #508.  This is a cleaner commit as well.

### CODE REVIEW

**Comment by bbooth** [comment]
Sorry for taking so long to get this back going.  Finally got some time to work on side projects again.  Let me know if there are anything you want changed.

**Comment by marcoow** [comment]
I think now that ESA will be only for Ember CLI it's better to include base64 as a bower dependency so we don't have to maintain it in the project itself - I think wrapping https://github.com/davidchambers/Base64.js as an Ember CLI addon (e.g. ember-cli-base64) would probably be cool anyway.

**Comment by bbooth** [comment]
I created a new ember-cli addon with the base64 polyfill, and updated the PR.  I will look into the merge conflicts now.

**Comment by marcoow** [comment]
👍

**Comment by bbooth** [comment]
Anything else I need to address?

File: addon/authenticators/oauth2.js [hunk 1]
```diff
@@ -223,13 +237,24 @@ export default Base.extend({
...
-    });
+      dataType:     'json',
+      contentType:  'application/x-www-form-urlencoded'
+    };
+
+    if (!Ember.isEmpty(this.clientId)) {
+      let base64ClientId = window.btoa(this.clientId.concat(':'));
+      Ember.merge(options, {
+        headers: {
+          Authorization: 'Basic '.concat(base64ClientId)
```
Comment (marcoow) [inline addon/authenticators/oauth2.js]: should be an ES2015 template string:

```js
Authorization: `Basic ${base64ClientId}`
```

**Comment by marcoow** [comment]
Looks good - just one small thing.

Can you also rebase on the `ember-cli-addon` branch please?

**Comment by bbooth** [comment]
Should be good to go.  Rebased and made the change to use a template string.  The ci tests are failing for ember-beta, but they don't seem to be an issue with my changes.  It seems to be the view deprecations  in 2.0

**Comment by marcoow** [comment]
👍

**Comment by csprocket777** [comment]
How do I implement this? Where in my config do I specify the client_id?

**Comment by bbooth** [comment]
With the latest release I specify the clientId in the custom authenticator.


/my-app/app/authenticatiors/oauth2.js
```
import OAuth2PasswordGrant from 'ember-simple-auth/authenticators/oauth2-password-grant';

export default OAuth2PasswordGrant.extend({
  clientId: 'my-client-id',
  serverTokenEndpoint: 'https://some-server/oauth/token'
});
```

**Comment by csprocket777** [comment]
Do you provide the client_secret the same way?

**Comment by bbooth** [comment]
You should not be providing a client secret in a Single Page Application, because it is insecure.  The addition of the clientId was mostly to support OAuth Servers that require client identification even if it is not secure.  Client secret was purposely left out.

**Comment by csprocket777** [comment]
Ok. Thank you, I need to figure out how to provide my backend with the client_secret then as it seems to require it.

**Comment by bbooth** [comment]
The only true way to do it securely is to proxy all oauth calls to a backend that knows both the client_id and client_secret.  Checkout this article.  It helped me a lot to understanding the issues associated with OAuth and javascript apps.  http://alexbilbie.com/2014/11/oauth-and-javascript/

**Pull request merged**